### PR TITLE
Check demultiplexed runs

### PR DESCRIPTION
**Pull request by barrystokman**
## Description

Script to clean up incorrect flowcells from demultiplexed-runs

### Added

`cg clean invalid-flowcell-dirs`

Directories of demultiplexed flowcells in `demultiplexed-runs` are checkd for correctness:
1. Name (removes renamed flowcells, often used to back up directories)
2. Exists in statusdb (a correctly demultiplexed flowcell is in post demux processing added to the Flowcell table in statusdb)
3. Fastq files in Housekeeper (a correctly demultiplexed flowcell has the fastq files added to Housekeeper in post demux processing)
4. Fastq files exist on hasta (check if fastq files in housekeeper actually exist on Hasta) 

`cg clean fix-flowcell-status`

Flowcell objects in statusdb are checked for the correct `status`:
1. A flowcell in statusdb that is present in `demultiplexed-runs` shoud have the status `ondisk`
2. Otherwise the flowcell should have the status `removed`
3. Flowcells that are being fetched from PDC, or that have been fetched but have not been demux yet are excluded from this check
  
### How to prepare for test
- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh [THIS-BRANCH-NAME]`

### Logical Test Cases:
1. Flowcell with with wrong name
2. Flowcell does not exists in statusdb
3. Flowcell with no fastq files in housekeeper
4. Flowcell fastq files in housekeeper, but not on disk
5. Flowcell is correct

6. Flowcell has status ondisk, is ondisk
7. Flowcell has status ondisk, is not ondisk
8. Flowcell has status removed, is ondisk
9. Flowcell has status removed, is not ondisk

### Test Data
1. 210712_A00689_0306_BHNVT7DSXY.bak
2. 211021_A00689_0375_AHLHGLDRXY
3. 211208_A00187_0648_BHMWNVDRXY
4. 211125_A00187_0632_AHLMHHDRXY
5. 211202_A00689_0415_AHLLHGDRXY
6. 211202_A00689_0415_AHLLHGDRXY (same as 5)
7. 211029_A00621_0522_AHHH5TDSX2
8. 211029_A00621_0521_BHLHLJDRXY
9. 210901_A00621_0482_AHJLTKDRXY

### Preconditions:
1. copy `210712_A00689_0306_BHNVT7DSXY` to ` 210712_A00689_0306_BHNVT7DSXY.bak`
2. ```DELETE FROM `cg-stage`.flowcell_sample where flowcell_id = 1608```, ```SELECT * FROM `cg-stage`.flowcell where name = 'HLHGLDRXY'```
3. `housekeeper delete files -t HMWNVDRXY --yes`
4. `rm -rf /home/proj/stage/demultiplexed-runs/211125_A00187_0632_AHLMHHDRXY/Unaligned`
5. N/A
6. N/A
7. `cg set flowcell HLHLJDRXY -s removed`
8. `cg set flowcell HJLTKDRXY -s removed`

### How to test
- [x] 1. `cg clean invalid-flowcell-dirs --dry-run`
- [ ] 2. `cg clean invalid-flowcell-dirs`
- [x] 3. `cg clean fix-flowcell-status --dry-run`
- [ ] 4. `cg clean fix-flowcell-status`

### Expected test outcome
- [x] 1 - 5:
```
Flowcell name not correctly named!: /home/proj/stage/demultiplexed-runs/210712_A00689_0306_BHNVT7DSXY.bak
Flowcell HNVT7DSXY failed one or more tests, setting flag to False!
Flowcell HLHGLDRXY does not exist in statusdb!
Flowcell HLHGLDRXY failed one or more tests, setting flag to False!
Flowcell HMWNVDRXY does not have any fastq files in Housekeeper!
Flowcell HMWNVDRXY has no fastq files in Housekeeper, skipping disk check
Flowcell HMWNVDRXY failed one or more tests, setting flag to False!
Flowcell HLMHHDRXYs has no fastq files on disk!
Flowcell HLMHHDRXY failed one or more tests, setting flag to False!
Flowcell HLLHGDRXY has passed all checks, setting flag to True!

═══════════════════════════════════╤═══════════════╤═════════════════╤═══════════════════════╤══════════════════════╤════════════════════════╤═════════════════╕
│ Flowcell name                     │ Flowcell id   │ Correct name?   │ Exists in statusdb?   │ Fastq files in HK?   │ Fastq files on disk?   │ Check passed?   │
├───────────────────────────────────┼───────────────┼─────────────────┼───────────────────────┼──────────────────────┼────────────────────────┼─────────────────┤
│ 210712_A00689_0306_BHNVT7DSXY.bak │ HNVT7DSXY     │ False           │ True                  │ True                 │ True                   │ False           │
╞═══════════════════════════════════╪═══════════════╪═════════════════╪═══════════════════════╪══════════════════════╪════════════════════════╪═════════════════╡
│ 211021_A00689_0375_AHLHGLDRXY     │ HLHGLDRXY     │ True            │ False                 │ True                 │ True                   │ False           │
├───────────────────────────────────┼───────────────┼─────────────────┼───────────────────────┼──────────────────────┼────────────────────────┼─────────────────┤
│ 211208_A00187_0648_BHMWNVDRXY     │ HMWNVDRXY     │ True            │ True                  │ False                │ False                  │ False           │
├───────────────────────────────────┼───────────────┼─────────────────┼───────────────────────┼──────────────────────┼────────────────────────┼─────────────────┤
│ 211125_A00187_0632_AHLMHHDRXY     │ HLMHHDRXY     │ True            │ True                  │ True                 │ False                  │ False           │
├───────────────────────────────────┼───────────────┼─────────────────┼───────────────────────┼──────────────────────┼────────────────────────┼─────────────────┤
│ 211202_A00689_0415_AHLLHGDRXY     │ HLLHGDRXY     │ True            │ True                  │ True                 │ True                   │ True            │
╘═══════════════════════════════════╧═══════════════╧═════════════════╧═══════════════════════╧══════════════════════╧════════════════════════╧═════════════════╛

Invalid flowcell directory found: /home/proj/stage/demultiplexed-runs/211021_A00689_0375_AHLHGLDRXY
Invalid flowcell directory found: /home/proj/stage/demultiplexed-runs/211125_A00187_0632_AHLMHHDRXY
Invalid flowcell directory found: /home/proj/stage/demultiplexed-runs/200910_A00689_0168_AHFLWJDSXY
Invalid flowcell directory found: /home/proj/stage/demultiplexed-runs/210712_A00689_0306_BHNVT7DSXY.bak
Invalid flowcell directory found: /home/proj/stage/demultiplexed-runs/211208_A00187_0648_BHMWNVDRXY
```
- [x] 6 - 9:
```
Setting status of flowcell HLHLJDRXY from removed to ondisk
Setting status of flowcell HHH5TDSX2 from ondisk to removed
```
- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch
- [ ] Inform to ...

### CODE REVIEW

**Comment by barrystokman** [comment]
TC1:
image
image

All tests pass, but there are some duplicate line `Flowcell <flowcell_id> has passed all checks, setting flag to True!` that I can't reproduce locally.

**Comment by barrystokman** [comment]
image
image
 ✔️

**Comment by barrystokman** [comment]
removed -> ondisk:
`2021-12-14 18:27:16 hasta.scilifelab.se cg.cli.clean[185633] INFO Setting status of flowcell HLHLJDRXY from removed to ondisk`
ondisk -> removed:
`2021-12-14 18:27:16 hasta.scilifelab.se cg.cli.clean[185633] INFO Setting status of flowcell HHH5TDSX2 from ondisk to removed`

File: cg/meta/clean/demultiplexed_flowcells.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                    self.exists_in_statusdb,
+                    self.fastq_files_exist_in_housekeeper,
+                    self.fastq_files_exist_on_disk,
+                ]
+            )
+            if self._passed_check:
+                LOG.info("Flowcell %s has passed all checks, setting flag to True!", self.id)
+        return self._passed_check
+
+    def check_existing_flowcell_directory(self):
```
Comment (patrikgrenfeldt) [inline cg/meta/clean/demultiplexed_flowcells.py:112]: Why do we need both this method and the 'passed_check' above? Seems like the purpose of this one is only to call the former one

File: tests/meta/clean/test_clean_demultiplexed_runs.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    ],
+)
+@mock.patch("cg.apps.housekeeper.hk.HousekeeperAPI")
+@mock.patch("cg.store.Store")
+def test_flowcell_name(
+    mock_statusdb: Store, mock_hk: HousekeeperAPI, flowcell_path: Path, result: bool, request
+):
+    # GIVEN a flowcell
+    flowcell_path = request.getfixturevalue(flowcell_path)
+    mock_hk.files.return_value.count.return_value = 1
```
Comment (patrikgrenfeldt) [inline tests/meta/clean/test_clean_demultiplexed_runs.py:30]: There's a danger that this deep mocking hinders changes to the implementation, see https://www.youtube.com/watch?v=qPfQM4w4I04

**Review by patrikgrenfeldt — APPROVED** [review]
Great work! I had some comments but nothing to hard to fix I hope 😄

File: cg/cli/clean.py [hunk 3]
```diff
@@ -177,3 +189,99 @@ def hk_bundle_files(
...
+    click.echo(
+        tabulate(
+            tabulate_row,
+            headers=FLOWCELL_OUTPUT_HEADERS,
+            tablefmt="fancy_grid",
+            missingval="N/A",
+        ),
+    )
+
+    failed_flowcells = [flowcell for flowcell in checked_flowcells if not flowcell.passed_check]
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/cli/clean.py:240]: If you agree to the suggestion above that is 😃

File: tests/meta/clean/conftest.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    """Full path to a correctly named flowcell directory in demultiplexed-runs"""
+    return Path(demultiplexed_flowcells_working_directory / correct_flowcell)
+
+
+@pytest.fixture(name="incorrect_flowcell_path_too_long")
+def fixture_incorrect_flowcell_path_too_long(
+    demultiplexed_flowcells_working_directory, incorrect_flowcell_too_long
+) -> Path:
+    """Full path to an incorrectly named flowcell directory in demultiplexed-runs"""
+    return Path(demultiplexed_flowcells_working_directory / incorrect_flowcell_too_long)
```
Comment (henrikstranneheim) [inline tests/meta/clean/conftest.py:44]:
```suggestion
    return Path(demultiplexed_flowcells_working_directory, incorrect_flowcell_too_long)
```
And also for the rest

**Review by henrikstranneheim — APPROVED** [review]
Very nice Barry!

**Comment by barrystokman** [comment]
@henrikstranneheim  I agree, I'll remove that part

**Comment by barrystokman** [comment]
retest:
image
image

I noticed an oddity here: flowcell `HLMHHDRXY` has no fastq files in Housekeeper, nor any spring files. This would tag this flowcell for deletion.
Running `housekeeper get file -t HLMHHDRXY -t fastq` confirms the absence of fastq files:
image
There are however fastq files in Housekeeper:
image
This is because the samples were sequenced and demultiplexed again at a later point in time, see flowcell `HLMV3DRXY`, which was sequenced a day later. A bundle already exists, the new fastq files get added to the bundle and then all the fastq files in the bundle get the tag of the newer flowcell, inclucing the older fastq files:
image
Notice how there's a fastq file `HLMHHDRXY_ACC9006A1_S3_L001_R1_001.fastq.gz` with the tag `HLMV3DRXY`.

When deleting flowcell `HLMHHDRXY`, the fastq files in HK will not be deleted since those files do not have the tag of that flowcell. Instead, they have the tag of the later flowcell.

This is slightly confusing and I'm not sure if this should be the intended behaviour,

**Comment by barrystokman** [comment]
There's two defects that need to be fixed in cg, but I have worked around them for now:

1. When adding samples to housekeeper in the post demultiplexing processes, and the sample already exists in a bundle, the older fastq files should NOT have their flowcell tag updated.
2. When spring compressing fastq files, the fastq files should NOT lose their flowcell tag

These fixes don't work retrospectively...

File: cg/cli/clean.py [hunk 5]
```diff
@@ -177,3 +191,118 @@ def hk_bundle_files(
...
+            tablefmt="fancy_grid",
+            missingval="N/A",
+        ),
+    )
+
+    for flow_cell in failed_flow_cells:
+        LOG.warning("Invalid flow cell directory found: %s", flow_cell.path)
+        if dry_run:
+            continue
+        LOG.warning("Removing %s!", flow_cell.path)
```
Comment (Mropat) [inline cg/cli/clean.py:256]:
```suggestion
        LOG.warning("Invalid flow cell directory found: %s", flow_cell.path)
        LOG.warning("Removing %s!", flow_cell.path)
        if dry_run:
            continue

```

**Review by Mropat — APPROVED** [review]

**Comment by barrystokman** [comment]
image

**Pull request merged**